README.md
# Format [![License][LicenseIMGURL]][LicenseURL] [![NPM version][NPMIMGURL]][NPMURL] [![Dependency Status][DependencyStatusIMGURL]][DependencyStatusURL] [![Build Status][BuildStatusIMGURL]][BuildStatusURL] [![Coverage Status][CoverageIMGURL]][CoverageURL]

Format size, permissions, etc.

# How to use?

Format could be used in browser or node.

```js
const format = require('format-io');
```

# API

## size

```js
const size = 1024 * 1024 * 5;
format.size(size);
// returns
'5.00mb'
```

## permissions.symbolic

```js
const perm = '00777';
format.permissions.symbolic(perm);
// returns
'rwx rwx rwx'
```

## permissions.numeric

```js
const perm = 'rwx rwx rwx';
format.permissions.numeric(perm);
// returns
'00777'
```

## Environments

In environments that not supports `es2015`, `format` can be used with:

```js
var format = require('format/legacy');
```

# License

MIT

[NPMIMGURL]:                https://img.shields.io/npm/v/format-io.svg?style=flat
[BuildStatusIMGURL]:        https://img.shields.io/travis/coderaiser/format-io/master.svg?style=flat
[DependencyStatusIMGURL]:   https://img.shields.io/david/coderaiser/format-io.svg?style=flat
[LicenseIMGURL]:            https://img.shields.io/badge/license-MIT-317BF9.svg?style=flat
[NPMURL]:                   https://npmjs.org/package/format-io "npm"
[BuildStatusURL]:           https://travis-ci.org/coderaiser/format-io  "Build Status"
[DependencyStatusURL]:      https://david-dm.org/coderaiser/format-io "Dependency Status"
[LicenseURL]:               https://tldrlegal.com/license/mit-license "MIT License"

[CoverageURL]:              https://coveralls.io/github/coderaiser/format-io?branch=master
[CoverageIMGURL]:           https://coveralls.io/repos/coderaiser/format-io/badge.svg?branch=master&service=github



ChangeLog
2018.10.10, v1.0.3

feature:
- (package) redrun v7.0.2
- (package) nyc v13.0.1
- (package) eslint-plugin-node v7.0.1
- (package) eslint v5.6.1
- (package) currify v3.0.0


2018.06.14, v1.0.2

feature:
- (mode) move out permissions


2018.06.12, v1.0.1

fix:
- (npmignore) legacy


2018.06.12, v1.0.0

feature:
- (format) es2015-ify


2015.02.18, v0.9.6

feature:
- (bower) add
- (format) scope -> global



test/format.js
'use strict';

const test = require('tape');
const {
    size,
    addSlashToEnd,
    permissions,
} = require('..');

test('format: addSlashToEnd: no path', (t) => {
    t.throws(addSlashToEnd, /path could not be empty!/, 'should throw');
    t.end();
});

test('format: addSlashToEnd: no slash', (t) => {
    const result = addSlashToEnd('hello');
    
    t.equal(result, 'hello/', 'should equal');
    t.end();
});

test('format: addSlashToEnd: slash', (t) => {
    const result = addSlashToEnd('hello/');
    
    t.equal(result, 'hello/', 'should equal');
    t.end();
});

test('format: size: not a number', (t) => {
    const result = size('-----');
    const expected = '-----';
    
    t.equal(result, expected, 'should equal');
    t.end();
});

test('format: size: b', (t) => {
    const result = size(10 ** 3);
    const expected = '1000b';
    
    t.equal(result, expected, 'should equal');
    t.end();
});

test('format: size: kb', (t) => {
    const result = size(10 ** 4);
    const expected = '9.77kb';
    
    t.equal(result, expected, 'should equal');
    t.end();
});

test('format: size: mb', (t) => {
    const result = size(10 ** 7);
    const expected = '9.54mb';
    
    t.equal(result, expected, 'should equal');
    t.end();
});

test('format: size: gb', (t) => {
    const result = size(10 ** 10);
    const expected = '9.31gb';
    
    t.equal(result, expected, 'should equal');
    t.end();
});

test('format: size: tb', (t) => {
    const result = size(10 ** 13);
    const expected = '9.09tb';
    
    t.equal(result, expected, 'should equal');
    t.end();
});

test('format: size: pb', (t) => {
    const result = size(10 ** 16);
    const expected = '8.88pb';
    
    t.equal(result, expected, 'should equal');
    t.end();
});

test('format: permissions: symbolic', (t) => {
    const perm = (16894).toString(8);
    const result = permissions.symbolic(perm);
    const expected = 'rwx rwx rw-';
    
    t.equal(result, expected, 'should equal');
    t.end();
});

test('format: permissions: symbolic: no args', (t) => {
    const result = permissions.symbolic();
    const expected = '';
    
    t.equal(result, expected, 'should equal');
    t.end();
});

test('format: permissions: numeric', (t) => {
    const result = permissions.numeric('rwx rwx rwx');
    const expected = '00777';
    
    t.equal(result, expected, 'should equal');
    t.end();
});

test('format: permissions: numeric: wrong format', (t) => {
    t.throws(permissions.numeric, /permissions should be in format "xxx xxx xxx"/, 'should equal');
    t.end();
});

test('format: permissions: numeric', (t) => {
    const result = permissions.numeric('rwx rwx rwx');
    const expected = '00777';
    
    t.equal(result, expected, 'should equal');
    t.end();
});



lib/mode.js
'use strict';

const currify = require('currify/legacy');

/*
    S_IRUSR   0000400   protection: readable by owner
    S_IWUSR   0000200   writable by owner
    S_IXUSR   0000100   executable by owner
    S_IRGRP   0000040   readable by group
    S_IWGRP   0000020   writable by group
    S_IXGRP   0000010   executable by group
    S_IROTH   0000004   readable by all
    S_IWOTH   0000002   writable by all
    S_IXOTH   0000001   executable by all
*/

const R = {
    name: 'r',
    value: 4,
};

const W = {
    name: 'w',
    value: 2,
};

const X = {
    name: 'x',
    value: 1,
};

const getModeName = currify((value, m) => {
    if (value & m.value)
        return m.name;
    
    return '-';
});

const toStrMode = currify((fn, value) => {
    return [R, W, X]
        .map(fn(value))
        .join('');
});


/**
 * Функция переводит права из цыфрового вида в символьный
 * @param perms - строка с правами доступа
 * к файлу в 8-миричной системе
 */
module.exports.symbolic = (perms) => {
    let permissions = '';
    const is = typeof perms !== 'undefined';
    
    if (!is)
        return permissions;
    
    const permsStr = perms.slice(-3);
    
    /* Переводим в двоичную систему */
    const owner = Number(permsStr[0]).toString(2);
    const group = Number(permsStr[1]).toString(2);
    const all   = Number(permsStr[2]).toString(2);
    
    const allPermissions = [
        owner,
        group,
        all,
    ];
    
    return allPermissions
        .map(toStrMode(getModeName))
        .join(' ');
};

/**
 * Функция конвертирует права доступа к файлам из символьного вида
 * в цыфровой
 */
module.exports.numeric = (perms) => {
    const length = perms && perms.length === 11;
    
    if (!length)
        throw Error('permissions should be in format "xxx xxx xxx"');
    
    const R = 4;
    const W = 2;
    const X = 1;
    const N = 0;
    
    const owner =
        (perms[0]  === 'r' ? R : N) +
        (perms[1]  === 'w' ? W : N) +
        (perms[2]  === 'x' ? X : N);
    
    const group =
        (perms[4]  === 'r' ? R : N) +
        (perms[5]  === 'w' ? W : N) +
        (perms[6]  === 'x' ? X : N);
     
    const all =
        (perms[8]  === 'r' ? R : N) +
        (perms[9]  === 'w' ? W : N) +
        (perms[10] === 'x' ? X : N);
    
    /* добавляем 2 цифры до 5 */
    return '00' + owner + group + all;
};



lib/format.js
'use strict';

module.exports.addSlashToEnd = (path) => {
    if (!path)
        throw Error('path could not be empty!');
    
    const length  = path.length - 1;
    const isSlash = path[length] === '/';
    
    if (isSlash)
        return path;
    
    return `${path}/`;
};

/** Функция получает короткие размеры
 * конвертируя байт в килобайты, мегабойты,
 * гигайбайты и терабайты
 * @pSize - размер в байтах
 */
module.exports.size = (size) => {
    const isNumber = typeof size === 'number';
    
    const l1KB = 1024;
    const l1MB = l1KB * l1KB;
    const l1GB = l1MB * l1KB;
    const l1TB = l1GB * l1KB;
    const l1PB = l1TB * l1KB;
    
    if (!isNumber)
        return size;
        
    if (size < l1KB)
        return size + 'b';
    
    if (size < l1MB)
        return (size / l1KB).toFixed(2) + 'kb';
    
    if (size < l1GB)
        return (size / l1MB).toFixed(2) + 'mb';
    
    if (size < l1TB)
        return (size / l1GB).toFixed(2) + 'gb';
    
    if (size < l1PB)
        return (size / l1TB).toFixed(2) + 'tb';
    
    return (size / l1PB).toFixed(2) + 'pb';
};

/**
 * Функция переводит права из цыфрового вида в символьный
 * @param perms - строка с правами доступа
 * к файлу в 8-миричной системе
 */
module.exports.permissions = {};
module.exports.permissions.symbolic = (perms) => {
    const is = typeof perms !== 'undefined';
    let permissions = '';
    /*
        S_IRUSR   0000400   protection: readable by owner
        S_IWUSR   0000200   writable by owner
        S_IXUSR   0000100   executable by owner
        S_IRGRP   0000040   readable by group
        S_IWGRP   0000020   writable by group
        S_IXGRP   0000010   executable by group
        S_IROTH   0000004   readable by all
        S_IWOTH   0000002   writable by all
        S_IXOTH   0000001   executable by all
    */
    
    if (!is)
        return permissions;
    
    const permsStr = perms.slice(-3);
    
    /* Переводим в двоичную систему */
    const owner = (permsStr[0] - 0).toString(2);
    const group = (permsStr[1] - 0).toString(2);
    const all   = (permsStr[2] - 0).toString(2);
    
    /* переводим в символьную систему*/
    permissions =
                 (owner[0] - 0 > 0 ? 'r' : '-')     +
                 (owner[1] - 0 > 0 ? 'w' : '-')     +
                 (owner[2] - 0 > 0 ? 'x' : '-')     +
                 ' '                                +
                 (group[0] - 0 > 0 ? 'r' : '-')     +
                 (group[1] - 0 > 0 ? 'w' : '-')     +
                 (group[2] - 0 > 0 ? 'x' : '-')     +
                 ' '                                +
                 (all[0] - 0   > 0 ? 'r' : '-')     +
                 (all[1] - 0   > 0 ? 'w' : '-')     +
                 (all[2] - 0   > 0 ? 'x' : '-');
    
    return permissions;
};

/**
 * Функция конвертирует права доступа к файлам из символьного вида
 * в цыфровой
 */
module.exports.permissions.numeric = (perms) => {
    const length = perms && perms.length === 11;
    
    if (!length)
        throw Error('permissions should be in format "xxx xxx xxx"');
    
    const R = 4;
    const W = 2;
    const X = 1;
    const N = 0;
    
    const owner =
        (perms[0]  === 'r' ? R : N) +
        (perms[1]  === 'w' ? W : N) +
        (perms[2]  === 'x' ? X : N);
    
    const group =
        (perms[4]  === 'r' ? R : N) +
        (perms[5]  === 'w' ? W : N) +
        (perms[6]  === 'x' ? X : N);
     
    const all =
        (perms[8]  === 'r' ? R : N) +
        (perms[9]  === 'w' ? W : N) +
        (perms[10] === 'x' ? X : N);
    
    /* добавляем 2 цифры до 5 */
    return '00' + owner + group + all;
};

